File: cses/02_sorting_searching/1631_reading-books/solution.py
import sys


def main():
    # Nạp TOÀN BỘ input một lần thành danh sách token (bộ nhớ O(n), không stream).
    data = sys.stdin.buffer.read().split()
    n = int(data[0])

    # Duyệt qua từng quyển sách để tính đồng thời:
    #   total = S = tổng thời gian tất cả các sách
    #   mx    = M = thời gian của quyển dài nhất
    total = 0
    mx = 0
    for i in range(1, n + 1):
        t = int(data[i])
        total += t
        if t > mx:
            mx = t

    # Đáp án = max(2*M, S): nếu quyển dài nhất quá lớn (2*M > S) thì bị nghẽn
    # bởi 2*M, ngược lại đủ chỗ xếp lịch song song nên bằng S.
    ans = max(2 * mx, total)
    sys.stdout.write(str(ans) + "\n")


main()

File: cses/02_sorting_searching/1631_reading-books/solution.cpp
#include <algorithm>
#include <iostream>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    // Duyệt qua từng quyển sách để tính đồng thời:
    //   total = S   = tổng thời gian tất cả các sách
    //   longest = M = thời gian của quyển dài nhất
    long long total = 0;
    long long longest = 0;
    for (int i = 0; i < n; ++i) {
        long long time;
        cin >> time;
        total += time;
        longest = max(longest, time);
    }

    // Đáp án = max(S, 2*M): nếu quyển dài nhất quá lớn (2*M > S) thì bị nghẽn
    // bởi 2*M, ngược lại đủ chỗ xếp lịch song song nên bằng S.
    cout << max(total, 2 * longest) << '\n';
    return 0;
}
